solver/src/driver.rs

```diff
@@ -125,38 +126,89 @@ impl Driver {
-                Ok(()) => Some(settlement),
-                Err(err) => {
-                    tracing::error!("settlement simulation failed\n error: {:?}", err);
-                    tracing::debug!("settlement failure for: \n{:#?}", settlement.settlement,);
-                    None
-                }
-            })
-            .collect())
+            .partition_map(|(settlement, result)| match result {
+                Ok(()) => Either::Left(settlement),
+                Err(err) => Either::Right((settlement, err)),
+            }))
+    }
+
+    // Log simulation errors only if the simulation also fails in the block at which on chain
+    // liquidity was queried. If the simulation succeeds at the previous block then the solver
+    // worked correctly and the error doesn't have to be reported.
+    // Note that we could still report a false positive because the earlier block might be off by if
+    // the block has changed just as were were querying the node.
+    async fn report_simulation_errors(
+        &self,
+        errors: Vec<(RatedSettlement, Error)>,
+        current_block_during_liquidity_fetch: u64,
+    ) {
+        let simulations = match settlement_simulation::simulate_settlements(
+            errors
+                .iter()
+                .map(|(settlement, _)| settlement.settlement.clone().into()),
+            &self.settlement_contract,
+            &self.web3,
+            &self.network_id,
+            settlement_simulation::Block::FixedWithTenderly(current_block_during_liquidity_fetch),
+        )
+        .await
+        {
+            Ok(simulations) => simulations,
+            Err(err) => {
+                tracing::error!(
+                    "failed to simulate failing settlements at earlier block {}: {:?}",
+                    current_block_during_liquidity_fetch,
+                    err
+                );
+                return;
+            }
+        };
+
+        for ((settlement, _previous_error), result) in errors.into_iter().zip(simulations) {
+            let error_at_earlier_block = match result {
+                Ok(()) => continue,
+                Err(err) => err,
+            };
+            tracing::error!(
+                "settlement simulation failed right before submission AND for block {} which was current when liquidity was fetched:\n{:?}",
+                current_block_during_liquidity_fetch, error_at_earlier_block
+            );
+            // This is an additional debug log so that the log message doesn't get too long as
+            // settlement information is recoverable through tenderly anyway.
+            tracing::debug!("settlement failure for: \n{:#?}", settlement.settlement,);
+        }
     }
 
     pub async fn single_run(&mut self) -> Result<()> {
         tracing::debug!("starting single run");
         let liquidity = self.liquidity_collector.get_liquidity().await?;
+        let current_block_during_liquidity_fetch = self
```

Comment:
I think it's an important follow up task to pass this block into the liquidity fetcher to make sure all pools are queried at the same block.